Add `spring-data-multi-storage-transaction-sample` project (#22)
scalar-labs/scalardb-samples: spring-data-multi-storage-transaction-sample/README.md

> # Spring Data Integration Sample with Multi-storage Transaction

Reviewer: ```suggestion
# Spring Data Integration Sample with Multi-storage Transactions
```

This tutorial describes how to create a sample Spring Boot application by using Spring Data integration with Multi-storage Transaction in ScalarDB.

## Prerequisites

- Java (OpenJDK 8 or higher)
- Gradle
- Docker, Docker Compose

## Sample application

### Overview

> This tutorial describes how to create a sample Spring Boot application for the same use case as [ScalarDB Sample](https://github.com/scalar-labs/scalardb-samples/tree/main/scalardb-sample) but by using Spring Data integration with Multi-storage Transaction in ScalarDB.
> Please note that application-specific error handling, authentication processing, etc. are omitted in the sample application since this tutorial focuses on explaining how to use Spring Data integration with Multi-storage Transaction in ScalarDB.
> For details, please see [Guide of Spring Data integration with ScalarDB](https://scalar-labs.github.io/scalardb-sql/spring-data-guide.html).

Reviewer: ```suggestion
This tutorial describes how to create a sample Spring Boot application for the same use case as [ScalarDB Sample](https://github.com/scalar-labs/scalardb-samples/tree/main/scalardb-sample) but by using Spring Data integration with Multi-storage Transactions in ScalarDB. Please note that application-specific error handling, authentication processing, etc. are omitted in the sample application since this tutorial focuses on explaining how to use Spring Data integration with Multi-storage Transactions in ScalarDB. For details, please see [Guide of Spring Data integration with ScalarDB](https://scalar-labs.github.io/scalardb-sql/spring-data-guide.html).
```

3. Getting order information by order ID
4. Getting order information by customer ID
5. Repayment (reduces the amount in the `credit_total`)

### Configuration

Configurations for the sample Spring Boot application are as follows:

```application.properties
spring.datasource.driver-class-name=com.scalar.db.sql.jdbc.SqlJdbcDriver
spring.datasource.url=jdbc:scalardb:\
?scalar.db.storage=multi-storage\
&scalar.db.multi_storage.storages=cassandra,mysql\
&scalar.db.multi_storage.storages.cassandra.storage=cassandra\
&scalar.db.multi_storage.storages.cassandra.contact_points=localhost\
&scalar.db.multi_storage.storages.cassandra.username=cassandra\
&scalar.db.multi_storage.storages.cassandra.password=cassandra\
&scalar.db.multi_storage.storages.mysql.storage=jdbc\
&scalar.db.multi_storage.storages.mysql.contact_points=jdbc:mysql://localhost:3306/\
&scalar.db.multi_storage.storages.mysql.username=root\
&scalar.db.multi_storage.storages.mysql.password=mysql\
&scalar.db.multi_storage.namespace_mapping=customer:mysql,order:cassandra,coordinator:cassandra\
&scalar.db.multi_storage.default_storage=cassandra
```

> - `scalar.db.storage`: `multi-storage` to use Multi-storage Transaction in ScalarDB

Reviewer: ```suggestion
- `scalar.db.storage`: Specifying `multi-storage` is necessary to use Multi-storage Transactions in ScalarDB.
```

> - `scalar.db.multi_storage.storages`: Definition of storage names

Reviewer: ```suggestion
- `scalar.db.multi_storage.storages`: The types of storage must be defined here.
```

> - `scalar.db.multi_storage.storages`: The types of storage must be defined here.

Author: @josh-wong On second thought, I think `names` is better than `types` since `cassandra` and `mysql` are not fixed values defined by ScalarDB but values users can freely define here (the values can be `kassandra,yoursql` here as long as the following properties use the same values.) `types` sounds pre-defined values to me. What do you think?

Reviewer: ```suggestion
- `scalar.db.multi_storage.storages`: Your storage names must be defined here.
```

Author: Thanks!

> - `scalar.db.multi_storage.storages.cassandra.*`: Configurations of `cassandra` storage that is one of the storages defined in `scalar.db.multi_storage.storages`. You can configure all the properties of `scalar.db.*` for the `cassandra` storage here

Reviewer: This was a bit tricky to reword. The word "storage" is similar to the word "data" in that, even if there are multiple storage units, the word doesn't become plural.

If the wording in my suggestion is incorrect, let's find a way to reword it!

```suggestion
- `scalar.db.multi_storage.storages.cassandra.*`: These configurations are for the `cassandra` storage, which is one of the storage types defined in `scalar.db.multi_storage.storages`. You can configure all the `scalar.db.*` properties for the `cassandra` storage here.
```

Author: Looks perfect to me. Thanks!!

> - `scalar.db.multi_storage.storages.cassandra.*`: These configurations are for the `cassandra` storage, which is one of the storage types defined in `scalar.db.multi_storage.storages`. You can configure all the `scalar.db.*` properties for the `cassandra` storage here.

Reviewer: ```suggestion
- `scalar.db.multi_storage.storages.cassandra.*`: These configurations are for the `cassandra` storage, which is one of the storage names defined in `scalar.db.multi_storage.storages`. You can configure all the `scalar.db.*` properties for the `cassandra` storage here.
```

> - `scalar.db.multi_storage.storages.mysql.*`: Configurations of `mysql` storage that is one of the storages defined in `scalar.db.multi_storage.storages`. You can configure all the properties of `scalar.db.*` for the `mysql` storage here

Reviewer: Similar to my earlier comment about the word "storage".

```suggestion
- `scalar.db.multi_storage.storages.mysql.*`: These configurations are for the `mysql` storage, which is one of the storage types defined in `scalar.db.multi_storage.storages`. You can configure all the `scalar.db.*` properties for the `mysql` storage here.
```

> - `scalar.db.multi_storage.storages.mysql.*`: These configurations are for the `mysql` storage, which is one of the storage types defined in `scalar.db.multi_storage.storages`. You can configure all the `scalar.db.*` properties for the `mysql` storage here.

Reviewer: ```suggestion
- `scalar.db.multi_storage.storages.mysql.*`: These configurations are for the `mysql` storage, which is one of the storage names defined in `scalar.db.multi_storage.storages`. You can configure all the `scalar.db.*` properties for the `mysql` storage here.
```

> - `scalar.db.multi_storage.namespace_mapping`: Mappings of namespace to storage. In this sample application, operations for `customer` namespace tables are mapped to the `mysql` storage and operations for `order` namespace tables are mapped to the `cassandra` storage. You can also define which storage is mapped for `coordinator` namespace that is used in Consensus Commit transaction

Reviewer: ```suggestion
- `scalar.db.multi_storage.namespace_mapping`: This configuration maps the namespaces to the storage. In this sample application, operations for `customer` namespace tables are mapped to the `mysql` storage and operations for `order` namespace tables are mapped to the `cassandra` storage. You can also define which storage is mapped for the `coordinator` namespace that is used in Consensus Commit transactions.
```

> - `scalar.db.multi_storage.default_storage`: Default storage that is used for operations on unmapped namespace tables

Reviewer: ```suggestion
- `scalar.db.multi_storage.default_storage`: This configuration sets the default storage that is used for operations on unmapped namespace tables.
```

> For details, please see [Multi-storage Transaction configurations](https://github.com/scalar-labs/scalardb/blob/master/docs/multi-storage-transactions.md).

Reviewer: ```suggestion
For details, please see [Configuration - Multi-storage Transactions](https://github.com/scalar-labs/scalardb/blob/master/docs/multi-storage-transactions.md#configuration).
```
